resolwe/flow/managers/listener/plugin.py

```diff
@@ -0,0 +1,125 @@
+"""Listener plugin and plugin manager class."""
+import abc
+from typing import (
+    Callable,
+    Dict,
+    Iterator,
+    MutableMapping,
+    Optional,
+    Set,
+    Type,
+)
+
+from resolwe.flow.executors.socket_utils import (
+    Message,
+    Response,
+)
+
+
+class ListenerPlugin(metaclass=abc.ABCMeta):
+    """Base class for listener plugin classes."""
+
+    @classmethod
+    def __init_subclass__(cls: Type["ListenerPlugin"], **kwargs):
+        """Register class with the registry on initialization."""
+        super().__init_subclass__(**kwargs)
+        ListenerPlugins().add_plugin(cls)
+
+    def _get_handler_name(self, command_name: str) -> str:
+        """Get the handler name."""
+        return f"handle_{command_name}"
+
+    def get_handler(
+        self, command_name: str
+    ) -> Optional[Callable[[Message, "Processor"], Response]]:
+        """Return the plugin handler
+
+        When no such plugin exists return None.
+        """
+        return getattr(self, self._get_handler_name(command_name), None)
+
+    def get_handlers(self) -> Set[str]:
+        """Return a list of handlers."""
+        return {
+            method_name
+            for method_name in dir(self)
+            if (
+                method_name.startswith("handle_")
+                and callable(getattr(self, method_name))
+            )
+        }
+
+
+class ListenerPlugins(MutableMapping[str, object]):
+    """Class for registering listener plugins."""
+
+    __instance = None  #  A single instance of this class
+    _plugin_classes: Dict[str, Type[ListenerPlugin]] = dict()
+    _plugins: Dict[str, ListenerPlugin] = dict()
+    _handler_names = set()
+
+    @classmethod
+    def __new__(cls, *args, **kwargs):
+        """Create or return singleton."""
+        if ListenerPlugins.__instance is None:
+            ListenerPlugins.__instance = super().__new__(cls)
+
+        return ListenerPlugins.__instance
+
+    def _get_class_name(self, plugin_class: Type[ListenerPlugin]) -> str:
+        """Get the full class name."""
+        return "{}.{}".format(plugin_class.__module__, plugin_class.__name__)
+
+    def add_plugin(self, plugin_class: Type[ListenerPlugin]):
+        """Add new plugin class.
+
+        :raises AssertionError: when plugin is already registered.
+        """
+        plugin_name = self._get_class_name(plugin_class)
+        assert (
+            plugin_name not in self._plugin_classes
+        ), "Plugin named {plugin_name} already registered."
+        self._plugin_classes[plugin_name] = plugin_class
+        plugin = plugin_class()
+        self._plugins[plugin_name] = plugin
```

Comment:
Add a debug log that the plugin was registered. I imagine it will come quite handy :).

Reply:
Sure.